packages/calling-stateful-client/src/DeviceManager.test.ts

```diff
@@ -0,0 +1,233 @@
+
+    expect(client.getState().deviceManager.selectedSpeaker).not.toBeDefined();
+
+    await (await client.getDeviceManager()).selectSpeaker(speaker);
+    expect(client.getState().deviceManager.selectedSpeaker).toBeDefined();
+    expect(client.getState().deviceManager.selectedSpeaker?.name).toBe('speaker');
+  });
+
+  test('should detect selectedSpeakerChanged update selectedSpeaker in state using deviceManager', async () => {
+    const manager = createMockDeviceManagerWithSpeakers([speakerWithName('speaker')]);
+    const client = createStatefulCallClientWithDeviceManager(manager);
+    // FIXME? Device manager is only created on request, so state is not updated unless device manager is created.
+    await client.getDeviceManager();
+
+    expect(client.getState().deviceManager.selectedSpeaker).not.toBeDefined();
+    manager.selectedSpeaker = speakerWithName('speaker');
+    manager.emit('selectedSpeakerChanged', {});
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.selectedSpeaker?.name === 'speaker')
+    ).toBe(true);
+  });
+
+  test('should detect audioDevicesUpdated update microphones/speakers in state using deviceManager', async () => {
+    const manager = createMockDeviceManagerWithSpeakers(
+      [speakerWithName('secondSpeaker')],
+      createMockDeviceManagerWithMicrophones([microphoneWithName('firstMicrophone')])
+    );
+    const client = createStatefulCallClientWithDeviceManager(manager);
+    // FIXME? Device manager is only created on request, so state is not updated unless device manager is created.
+    await client.getDeviceManager();
+
+    manager.emit('audioDevicesUpdated', {});
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.microphones[0]?.name === 'firstMicrophone')
+    ).toBe(true);
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.speakers[0]?.name === 'secondSpeaker')
+    ).toBe(true);
+  });
+
+  test('should proxy askDevicePermission and update deviceAccess in state', async () => {
+    const manager = addMockEmitter({
+      async askDevicePermission(): Promise<DeviceAccess> {
+        return {
+          audio: true,
+          video: true
+        };
+      }
+    }) as MockDeviceManager;
+    const client = createStatefulCallClientWithDeviceManager(manager);
+
+    const deviceAccess = await (await client.getDeviceManager()).askDevicePermission({ audio: false, video: false });
+    expect(deviceAccess.audio).toBe(true);
+    expect(deviceAccess.video).toBe(true);
+    expect(client.getState().deviceManager.deviceAccess).toBeDefined();
+    expect(client.getState().deviceManager.deviceAccess?.audio).toBe(true);
+    expect(client.getState().deviceManager.deviceAccess?.video).toBe(true);
+  });
+});
+
+const createStatefulCallClientWithDeviceManager = (deviceManager: MockDeviceManager): StatefulCallClient => {
```

Comment:
Does it make sense to move here and below to a helper file? (e.g. `DeviceManagerTestUtils` or something?) to help reduce the size of the file

Reply:
Don't have strong opinion on this.

Moving functions around is easy :)